ops/terraform/modules/resources/bfd_server_slo_alarms/main.tf

```diff
@@ -3,9 +3,47 @@ locals {
 
   app = "bfd-server"
 
-  alert_arn   = var.alert_notification_arn == null ? [] : [var.alert_notification_arn]
-  warning_arn = var.warning_notification_arn == null ? [] : [var.warning_notification_arn]
-  ok_arn      = var.ok_notification_arn == null ? [] : [var.ok_notification_arn]
+  victor_ops_sns         = "bfd-${var.env}-cloudwatch-alarms"
+  bfd_test_slack_sns     = "bfd-${var.env}-cloudwatch-alarms-slack-bfd-test"
+  bfd_warnings_slack_sns = "bfd-${var.env}-cloudwatch-alarms-slack-bfd-warnings"
+  default_ok_sns         = "bfd-${var.env}-cloudwatch-ok"
+  # Each established environment has a different destination for which alarm notifications should
+  # route to. The below map maps each particular SNS (destination) to a particular type of SLO
+  # alarm.
+  topic_names_by_env = {
+    prod = {
+      alert   = local.victor_ops_sns
+      warning = local.bfd_warnings_slack_sns
+      ok      = local.default_ok_sns
+    }
+    prod-sbx = {
+      alert   = local.bfd_test_slack_sns
+      warning = local.bfd_test_slack_sns
+      ok      = null
+    }
+    test = {
+      alert   = local.bfd_test_slack_sns
+      warning = local.bfd_test_slack_sns
+      ok      = null
+    }
+  }
+  # In the event this module is being applied in a non-established environment (i.e. an ephemeral
+  # environment) this lookup will ensure that an empty configuration will be returned
+  env_sns = lookup(local.topic_names_by_env, var.env, {
+    alert   = null
+    warning = null
+    ok      = null
+  })
+  # The following trys and coalesces ensure two things: the operator is able to override the
+  # SNS topic/destination of each alarm type, and that if no destination is specified (either
+  # explicitly such as with the OK SNS topics in prod-sbx/test or through the environment being
+  # ephemeral) that Terraform does not raise an error and instead the SNS topic is empty
+  alert_sns_name   = try(coalesce(var.alert_sns_override, local.env_sns.alert), null)
+  warning_sns_name = try(coalesce(var.warning_sns_override, local.env_sns.warning), null)
+  ok_sns_name      = try(coalesce(var.ok_sns_override, local.env_sns.ok), null)
+  alert_arn        = try([data.aws_sns_topic.alert_sns[0].arn], [])
+  warning_arn      = try([data.aws_sns_topic.warning_sns[0].arn], [])
+  ok_arn           = try([data.aws_sns_topic.ok_sns[0].arn], [])
```

Comment:
Interesting. I've never had a case for using `try` in vanilla terraform. Hmm.

Reply:
I'd be interested to know if there's a more "Terraformic" (i.e. _Pythonic_) way of doing the operations that are being done here. Basically, `coalesce` will error if all of its parameters evaluate to `null`, so I need to wrap that in `try` in order to ensure that if they're all `null` `null` is returned instead of an error. The second case usage of `try` here is _basically_ the equivalent of the `?.[n]` operator or wrapping the array with a `if array.length > 0` or something similar. I'm not sure if there's the equivalent of `map`'s `lookup` for arrays in Terraform, but if I'm wrong then that might be a better approach.

Reply:
The future's bright. Thanks for the thorough explanation!